chia/util/augmented_chain.py

```diff
@@ -0,0 +1,131 @@
+        # generators
+        to_remove = []
+        curr: Optional[Tuple[FullBlock, BlockRecord]] = self._extra_blocks.get(header_hash)
+        while curr is not None:
+            b = curr[0]
+            if b.height in generator_refs:
+                if b.transactions_generator is None:
+                    raise ValueError(Err.GENERATOR_REF_HAS_NO_GENERATOR)
+                generators[b.height] = bytes(b.transactions_generator)
+                to_remove.append(b.height)
+            header_hash = b.prev_header_hash
+            curr = self._extra_blocks.get(header_hash)
+        for i in to_remove:
+            generator_refs.remove(i)
+
+        if len(generator_refs) > 0:
+            generators.update(await self._underlying.lookup_block_generators(header_hash, generator_refs))
+        return generators
+
+    async def get_block_record_from_db(self, header_hash: bytes32) -> Optional[BlockRecord]:
+        ret = await self._underlying.get_block_record_from_db(header_hash)
+        if ret is not None:
+            return ret
+        return maybe_record(self._extra_blocks.get(header_hash))
+
+    def add_block_record(self, block_record: BlockRecord) -> None:
+        self._underlying.add_block_record(block_record)
+
+        # now that we're adding the block to the underlying blockchain, we don't
+        # need to keep the extra block around anymore
+        hh = block_record.header_hash
+        if hh in self._extra_blocks:
+            del self._height_to_hash[block_record.height]
+            del self._extra_blocks[hh]
+
+    # BlockRecordsProtocol
+    def try_block_record(self, header_hash: bytes32) -> Optional[BlockRecord]:
+        ret = self._underlying.try_block_record(header_hash)
+        if ret is None:
+            ret = maybe_record(self._extra_blocks.get(header_hash))
+        return ret
+
+    def block_record(self, header_hash: bytes32) -> BlockRecord:
+        ret = self._underlying.try_block_record(header_hash)
+        if ret is not None:
+            return ret
+        return self._extra_blocks[header_hash][1]
+
+    def height_to_block_record(self, height: uint32) -> BlockRecord:
+        header_hash = self._underlying.height_to_hash(height)
+        if header_hash is not None:
+            return self._underlying.block_record(header_hash)
+        header_hash = self._height_to_hash.get(height)
+        if header_hash is None:
+            raise ValueError(f"Height is not in blockchain: {height}")
+        return self._extra_blocks[header_hash][1]
+
+    def height_to_hash(self, height: uint32) -> Optional[bytes32]:
+        header_hash = self._underlying.height_to_hash(height)
+        if header_hash is not None:
+            return header_hash
+        return self._height_to_hash.get(height)
+
+    def contains_block(self, header_hash: bytes32) -> bool:
+        return (header_hash in self._extra_blocks) or self._underlying.contains_block(header_hash)
+
+    def contains_height(self, height: uint32) -> bool:
```

Comment:
I'm not a fan of this interface (or name) either, partly because it's a bit ambiguous. My understanding is that this checks if `height` is in the block record *cache*, not just whether the peak height > `height`.

Reply:
no this checks the height_to_hash not the cache i.e if < peak.height should be true